examples: Add shallow water equations notebook

AtilaSaraiva opened the pull request:
Hello folks, I've been working on a simple notebook which implements the 2D Shallow Water equations, which is generally used to model Tsunamis. I've made it while learning the Devito API, and hope that it might be useful to someone else.

EdCaunt commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>A few comments:</p><p> </p><ul><li>Possibly replace the hand-drawn wave and seabed lines with some kind of smooth curve? I think Bezier curves or similar would look a lot neater and cleaner. Really good figure otherwise.</li><li>Do either of the links you refer to have a DOI?</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

AtilaSaraiva replied:
<p>Thx for the review, first of all! I agree, I will make a better drawing on inkscape and update the notebook. About the DOI, not that I'm aware of.</p>

EdCaunt replied:
Is there any paper which outlines them or uses a comparable implementation (even if it's more complex than the examples shown here)? Not sure how much anyone else is bothered, but I think it would be good to be able to link to something with a DOI (so we don't get any broken links down the line)

AtilaSaraiva replied:
<p>Well, I could use an internet archive link, what do u think?</p>

EdCaunt replied:
Tbh, if there isn't an easy replacement than I'd just let it slide and change it if someone complains that the link is dead later., assuming no-one else is bothered by that

EdCaunt commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #5.    def op_shallowWaterWave(etasave, eta, M, N, h, D, g, alpha):
</pre><p>Do we use camel case for functions in Devito? I think <code>op_shallowwaterwave</code> would be more consistent with the style of the rest of the codebase. Double check this though </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

georgebisbas replied:
<p>I think we use </p><pre class="ql-syntax" spellcheck="false">def ForwardOperator()

</pre>

AtilaSaraiva replied:
<p>Hm, should I remove the underline? I'm not sure, but the camel case makes sense.</p>

EdCaunt commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #13.        eta : numpy.ndarray
</pre><p>Should these all be Devito <code>Function</code>s? The code implies they are  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

georgebisbas replied:
<p>Should typically be </p><p><code>ndarray or Function</code></p>

AtilaSaraiva replied:
<p>Oh sorry, I forgot to adapt it when I rewrote the function, they all are devito Functions yes.</p>

EdCaunt commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    print(optime.ccode)
</pre><p>Possibly comment this and let the user uncomment if they want, as the C code generated is fairly long and breaks up flow of notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

mloubout replied:
<p>I would agree with that.</p>

georgebisbas replied:
<p>I agree as well.</p><p>Something along the lines of</p><pre class="ql-syntax" spellcheck="false"># If you want to have a look at the generated code....uncomment...
#print(op.ccode)
</pre>

AtilaSaraiva replied:
<p>I do agree, makes sense, thx for the review!</p>

EdCaunt commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #8.    r = 2.0 * (np.random.rand(ny, nx) - 0.5) * pert # create random number perturbations
</pre><p>Probably a good idea to set a seed here for reproducibility</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

AtilaSaraiva replied:
<p>You're right, I should've thought about it.</p>

EdCaunt replied:
Will definitely need doing for the notebook CI to pass

EdCaunt commented:
Really cool notebook btw, it's a nice use case that we don't yet have any demos for

mloubout commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>The wave disapear in the middle of the simulation, is there a stability issue there or is it just becoming zeros?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

georgebisbas replied:
<p>I think it not capturing full timesteps in the video?Atila could you double check?</p><p>It seems like at the last frames a new blue trace is starting at the center</p>

AtilaSaraiva replied:
<p>Thx for the review! I actually pondered about that, I'm going to have to investigate, but indeed the video is capturing what is present in the etasave function, so I might have to review how it is capturing snapshots.</p>

mloubout commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>Looks like there is a stability issue too</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

georgebisbas replied:
<pre class="ql-syntax" spellcheck="false">Line #21.        D : numpy.ndarray
</pre><p>All numpy.ndarray -&gt; ndarray</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

georgebisbas replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    print(optime.ccode)
</pre><p>optime can be renamed as op</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

georgebisbas replied:
<p>Tanh -&gt; $tanh$</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

georgebisbas replied:
<p>... -&gt; :</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

mloubout commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>boring -&gt; simple?</p><p> </p><p>Maybe plot the that <code>tanh</code> function to see what the ocean floor looks like? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

AtilaSaraiva replied:
<p>I agree, it would be nice, gonna do that!</p>

mloubout commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>Any intuition/ref on how to choose that dt?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

AtilaSaraiva replied:
<p>I actually just copied from the original notebook, as I was trying to reproduce the same results, but from what I could gather the idea is maintaining a factor around 300.</p>

mloubout commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>Bit wordy but does explain the zeros. Any chance can truncate the snapshot so that it doesn't look like the wave disappear?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

AtilaSaraiva replied:
<p>Yeah, I can do that, but still, the first example shouldnt have this problem at all, so I still have some things to figure out.</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>devito -&gt; Devito</p><p>Tsunamis -&gt; Tsunami events?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

AtilaSaraiva replied:
<p>I agree!</p>

georgebisbas replied:
<p>Is this fixed?</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>we want to "model" the following problem?</p><p>In Eq. 1?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

AtilaSaraiva replied:
<p>Oh indeed, I will correct those.</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #9.    # Maximum wave propagation time [s]
</pre><p>Just add an empty line for readability</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

AtilaSaraiva replied:
<p>Agreed.</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #3.    # Defining devito symbolic functions
</pre><p>Can drop devito</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

AtilaSaraiva replied:
<p>Agreed too!</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #26.    optime = op_shallowWaterWave(etasave, eta, M, N, h, D, g, alpha)
</pre><p>optime can be just op</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

AtilaSaraiva replied:
<p>Yeah, better.</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>in an ocean -&gt; in ocean ?</p><p>Drop three dots at the end?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

AtilaSaraiva replied:
<p>Indeed</p>

georgebisbas commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,4210 @@
+{
```
<p>is copy() needed?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

AtilaSaraiva replied:
<p>I just wanted to garantee a copy by value and not by reference but indeed it isn't needed.</p>

AtilaSaraiva commented:
Guys, sorry for taking too long to take a look at your reviews, I've been quite these last two weeks with a paper, but I gonna start now!

AtilaSaraiva commented:
> Really cool notebook btw, it's a nice use case that we don't yet have any demos for

Oh thx! I'm happy that the notebook I've made to learn Devito is useful.

FabioLuporini commented:
Any updates on this @AtilaSaraiva @EdCaunt @georgebisbas (@mloubout) ?

AtilaSaraiva commented:
Added a commit addressing most of the comments, except from some open ended. I managed to find a quick and dirty fix for the stability issue doing several modifications to the code.

AtilaSaraiva commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,9323 @@
+{
```
Sorry, indeed a silly mistake. Fixed!

mloubout commented on examples/cfd/08_shallow_water_equation.ipynb:
```diff
@@ -0,0 +1,8079 @@
+{
```
<p>Maybe remove "initial" from M/N since they are time varying</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

AtilaSaraiva replied:
<p>Thanks for noticing, fixed!</p>

mloubout commented:
Needs rebase and GTG

AtilaSaraiva commented:
I thought I did that haha, lemme rebase it again

AtilaSaraiva commented:
Done!

mloubout commented:
Looks like some cells need to be fixed for CI and will be GTG

AtilaSaraiva commented:
oops I messed up, wait a sec

AtilaSaraiva commented:
omg please forgive my noobness, I'm trying to fix the tree

AtilaSaraiva commented:
Done, fixed, sorry for the delay.

mloubout commented:
The notebook is still not fix for CI. Please add the nbal tags to each cell. And run the test locally to make sure everything is working correctly before pushing. And please make sure you are fully rebased on top of the master branch.

`py.test --nbval  examples/cfd/08_shallow_water_equation.ipynb`

AtilaSaraiva commented:
okay I will look into it now.

AtilaSaraiva commented:
Now I think we're golden.
```
==================== test session starts ====================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0
rootdir: /home/atila/Files/Códigos/gitRepos/devito, configfile: setup.cfg
plugins: nbval-0.9.6
collected 25 items                                          

examples/cfd/08_shallow_water_equation.ipynb .....s.. [ 32%]
.s...s...s...s..s                                    [100%]

===================== warnings summary =====================
../../../../../../nix/store/725hmi98j8xnikz6fkx6wak1kfbfnj33-python3-3.9.12-env/lib/python3.9/site-packages/nbval/plugin.py:206
  /nix/store/725hmi98j8xnikz6fkx6wak1kfbfnj33-python3-3.9.12-env/lib/python3.9/site-packages/nbval/plugin.py:206: PytestRemovedIn8Warning: The (fspath: py.path.local) argument to IPyNbFile is deprecated. Please use the (path: pathlib.Path) argument instead.
  See https://docs.pytest.org/en/latest/deprecations.html#fspath-argument-for-node-constructors-replaced-with-pathlib-path
    super(IPyNbFile, self).__init__(*args, **kwargs)

examples/cfd/08_shallow_water_equation.ipynb::Cell 0
  /home/atila/Files/Códigos/gitRepos/devito/devito/types/basic.py:613: DeprecationWarning: invalid escape sequence \-
    """

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================== slowest 20 durations ===================
40.98s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 8
23.30s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 16
20.31s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 4
17.81s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 20
17.65s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 23
13.46s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 12
4.68s setup    examples/cfd/08_shallow_water_equation.ipynb::Cell 0
1.96s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 0
0.35s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 11
0.28s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 15
0.27s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 19
0.21s teardown examples/cfd/08_shallow_water_equation.ipynb::Cell 24
0.20s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 18
0.10s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 3
0.03s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 10
0.03s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 22
0.03s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 14
0.02s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 7
0.02s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 1
0.01s call     examples/cfd/08_shallow_water_equation.ipynb::Cell 2
```
